File: src/main/java/org/tobi29/scapes/entity/client/MobClient.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.entity.client;

import org.tobi29.scapes.chunk.WorldClient;
import org.tobi29.scapes.engine.server.PlayConnection;
import org.tobi29.scapes.engine.utils.io.tag.TagStructure;
import org.tobi29.scapes.engine.utils.math.AABB;
import org.tobi29.scapes.engine.utils.math.FastMath;
import org.tobi29.scapes.engine.utils.math.vector.MutableVector3d;
import org.tobi29.scapes.engine.utils.math.vector.Vector3;
import org.tobi29.scapes.entity.MobPositionHandler;
import org.tobi29.scapes.entity.MobileEntity;
import org.tobi29.scapes.entity.model.MobModel;

import java.util.Optional;

public abstract class MobClient extends EntityClient implements MobileEntity {
    protected final AABB collision;
    protected final MutableVector3d speed, rot = new MutableVector3d();
    protected final MobPositionHandler positionHandler;
    protected boolean ground, inWater, swimming, headInWater, slidingWall;

    protected MobClient(WorldClient world, Vector3 pos, Vector3 speed,
            AABB aabb) {
        super(world, pos);
        this.speed = new MutableVector3d(speed);
        collision = aabb;
        positionHandler = createPositionHandler(world.connection());
    }

    public AABB aabb() {
        AABB aabb = new AABB(collision);
        aabb.add(pos.doubleX(), pos.doubleY(), pos.doubleZ());
        return aabb;
    }

    @Override
    public void read(TagStructure tagStructure) {
        super.read(tagStructure);
        positionHandler.receiveMoveAbsolute(pos.doubleX(), pos.doubleY(),
                pos.doubleZ());
        tagStructure.getMultiTag("Speed", speed);
        tagStructure.getMultiTag("Rot", rot);
    }

    @Override
    public Optional<? extends MobModel> createModel() {
        return Optional.empty();
    }

    protected MobPositionHandler createPositionHandler(
            PlayConnection connection) {
        return new MobPositionHandler(pos.now(), connection::send, pos::set,
                speed::set, rot::set,
                (ground, slidingWall, inWater, swimming) -> {
                    this.ground = ground;
                    this.slidingWall = slidingWall;
                    this.inWater = inWater;
                    this.swimming = swimming;
                });
    }

    @Override
    public MobPositionHandler positionHandler() {
        return positionHandler;
    }

    public Vector3 speed() {
        return speed.now();
    }

    public Vector3 rot() {
        return rot.now();
    }

    public double speedX() {
        return speed.doubleX();
    }

    public double pitch() {
        return rot.doubleX();
    }

    public double speedY() {
        return speed.doubleY();
    }

    public double tilt() {
        return rot.doubleY();
    }

    public double speedZ() {
        return speed.doubleZ();
    }

    public double yaw() {
        return rot.doubleZ();
    }

    public boolean isHeadInWater() {
        return headInWater;
    }

    public boolean isInWater() {
        return inWater;
    }

    public boolean isSwimming() {
        return swimming;
    }

    public boolean isOnGround() {
        return ground;
    }

    public void move(double delta) {
        headInWater = world.terrain().type(pos.intX(), pos.intY(),
                FastMath.floor(pos.doubleZ() + 0.7)).isLiquid();
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/entity/client/MobClient.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.entity.client

import org.tobi29.scapes.chunk.WorldClient
import org.tobi29.math.AABB
import org.tobi29.math.vector.MutableVector3d
import org.tobi29.math.vector.Vector3d
import org.tobi29.scapes.entity.EntityPhysics
import org.tobi29.scapes.entity.EntityType
import org.tobi29.scapes.entity.Mob
import org.tobi29.scapes.entity.MobPositionReceiver
import org.tobi29.io.tag.TagMap
import org.tobi29.io.tag.toMap
import org.tobi29.stdex.math.floorToInt

abstract class MobClient(type: EntityType<*, *>,
                         world: WorldClient,
                         pos: Vector3d,
                         speed: Vector3d,
                         protected val collision: AABB) : EntityAbstractClient(
        type, world, pos),
        Mob,
        MobileEntityClient {
    protected val speed: MutableVector3d
    protected val rot = MutableVector3d()
    override val positionReceiver: MobPositionReceiver
    val isOnGround: Boolean
        get() = physicsState.isOnGround
    val isInWater: Boolean
        get() = physicsState.isInWater
    val isSwimming: Boolean
        get() = physicsState.isSwimming
    var isHeadInWater = false
        protected set
    protected var physicsState = EntityPhysics.PhysicsState()

    init {
        this.speed = MutableVector3d(speed)
        positionReceiver = MobPositionReceiver(this.pos.now(),
                { this.pos.set(it) },
                { this.speed.set(it) },
                { this.rot.set(it) },
                { ground, slidingWall, inWater, swimming ->
                    physicsState.isOnGround = ground
                    physicsState.slidingWall = slidingWall
                    physicsState.isInWater = inWater
                    physicsState.isSwimming = swimming
                })
    }

    override fun getAABB(): AABB {
        val aabb = AABB(collision)
        aabb.add(pos.x, pos.y, pos.z)
        return aabb
    }

    override fun read(map: TagMap) {
        super.read(map)
        positionReceiver.receiveMoveAbsolute(pos.x, pos.y,
                pos.z)
        map["Speed"]?.toMap()?.let { speed.set(it) }
        map["Rot"]?.toMap()?.let { rot.set(it) }
    }

    fun speed(): Vector3d {
        return speed.now()
    }

    fun rot(): Vector3d {
        return rot.now()
    }

    fun speedX(): Double {
        return speed.x
    }

    fun pitch(): Double {
        return rot.x
    }

    fun speedY(): Double {
        return speed.y
    }

    fun speedZ(): Double {
        return speed.z
    }

    fun yaw(): Double {
        return rot.z
    }

    open fun move(delta: Double) {
        isHeadInWater = world.terrain.type(pos.x.floorToInt(),
                pos.y.floorToInt(),
                (pos.z + 0.7).floorToInt()).isLiquid
    }
}
